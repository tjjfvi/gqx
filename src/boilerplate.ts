// @ts-nocheck
throw new Error("boilerplate.ts should not be imported")
/* eslint-disable @typescript-eslint/no-empty-function */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable no-unused-vars */

/* --- */

type Int = number;
type Float = number;
type String = string;
type ID = string;
type Boolean = boolean;

interface $<F, L> { f: F; l: L }
const $$ = <F, L>(f: F, l: L) => ({ f, l });

interface $$OperationId {
  typeProp: string;
  type: string;
  prop: string;
  inputTypes: { [x: string]: string };
}

const __wrap__ = Symbol();

type $$Wrap<X, Y> = X extends $<infer F, infer L> ? $<F, $$Wrap<L, Y>> : never;
type $$MapWrap<O, F> = {
  [K in keyof O | typeof __wrap__ | "$"]:
    K extends "$" ?
      <T extends $_>(x: T[]) => $<F, O extends { [__wrap__]: infer X } ? $$Wrap<X, T> : T>[] :
      K extends keyof O ?
        O[K] extends $_ ?
          $<F, O[K]> :
          $$MapWrap<O[K], F> :
        $<F, null>
}

const $$mapWrap = <O, F>(o: () => O, f: F): $$MapWrap<O, F> =>
  // @ts-ignore
  new Proxy(Object.create(null), {
    get: (t, k: string) => t[k] || (t[k] =
      k !== "$" ?
        o()[k].$ ?
          $$mapWrap(() => o()[k], f) :
          $$(f, o()[k]) :
        // @ts-ignore
        (a: any) => ("$" in o() ? o().$(a) : a).map((a: any) => $$(f, a))
    ),
  })

const $$reconstruct = <I extends $$OperationId>(id: I, input: $$Input.$<I>, props: $$Frag.$<I>[]) => {
  interface Subs { [k: string]: true | Subs }
  const subs: Subs = {};
  props.map(prop => populateSubs(prop, subs));
  const frag = genFrag(subs);
  const inputKeys = Object.keys(input);
  const inputDef = inputKeys.length ? `(${inputKeys.map(k =>
    `$${k}: ${id.inputTypes[k]}`
  ).join(", ")})` : "";
  const inputPass = inputKeys.length ? `(${inputKeys.map(k =>
    `${k}: $${k}`
  )})` : "";
  const request = `${id.typeProp}${inputDef} { ${id.prop}${inputPass}${frag} }`;
  return request;

  function genFrag(subs: Subs){
    return (
      Object.keys(subs).length ?
        `{ ${Object.entries(subs).map(([k, v]) => v === true ? k : k + " " + genFrag(v)).join(" ")} }` :
        ""
    );
  }

  function populateSubs(prop: $_, subs: Subs | true){
    if(subs === true)
      return;
    if("prop" in prop)
      subs[prop.prop] = true;
    else
      populateSubs(prop.l, subs[prop.f.prop] = (subs[prop.f.prop] || {}));
  }
}
