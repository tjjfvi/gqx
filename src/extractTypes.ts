import { DocumentNode, TypeNode } from "graphql"
import { Context } from "vm"
import { InputField, NamedType, Type, UnresolvedType } from "./type"

const baseScalars = ["ID", "String", "Int", "Float", "Boolean"]

export function extractTypes(documents: DocumentNode[]) {
  const types = new Map<string, NamedType>()

  for (const name of baseScalars) {
    types.set(name, {
      kind: "scalar",
      name,
      directives: [],
    })
  }

  const definitions = documents.flatMap(x => x.definitions)

  for (const typeDef of definitions) {
    if (typeDef.kind === "SchemaDefinition" || typeDef.kind === "SchemaExtension") {
      for (const operation of typeDef.operationTypes ?? []) {
        const type = getNamedType(operation.type.name.value)
        setKind(type, "object", { fields: [], subtypes: [], supertypes: [] })
        type.operationKey = operation.operation
      }
    }

    if (!("name" in typeDef) || !typeDef.name) continue

    const name = typeDef.name.value
    const type = getNamedType(name)

    if ("directives" in typeDef && typeDef.directives) {
      type.directives.push(...typeDef.directives)
    }

    if (typeDef.kind === "ScalarTypeDefinition" || typeDef.kind === "ScalarTypeExtension") {
      setKind(type, "scalar", {})
    } else if (
      typeDef.kind === "InputObjectTypeDefinition"
      || typeDef.kind === "InputObjectTypeExtension"
    ) {
      setKind(type, "input", { fields: [] })
      for (const fieldDef of typeDef.fields ?? []) {
        const key = fieldDef.name.value
        const fieldType = getType(fieldDef.type)
        type.fields.push({
          key,
          type: fieldType,
          directives: fieldDef.directives?.slice() ?? [],
        })
      }
    } else if (
      typeDef.kind === "ObjectTypeDefinition"
      || typeDef.kind === "ObjectTypeExtension"
      || typeDef.kind === "InterfaceTypeDefinition"
      || typeDef.kind === "InterfaceTypeExtension"
    ) {
      setKind(type, "object", { fields: [], subtypes: [], supertypes: [] })
      for (const fieldDef of typeDef.fields ?? []) {
        const key = fieldDef.name.value
        const fullType = getType(fieldDef.type)
        const namedType = getNamedTypeOfType(fullType)
        const name = `${type.name}$${key}`
        type.fields.push({
          str: JSON.stringify(name),
          name,
          base: type,
          key,
          namedType,
          fullType,
          directives: fieldDef.directives?.slice() ?? [],
          input: fieldDef.arguments?.map((arg): InputField => ({
            key: arg.name.value,
            type: getType(arg.type),
            directives: arg.directives?.slice() ?? [],
          })) ?? [],
        })
      }
      if ("interfaces" in typeDef) {
        for (const supertypeDef of typeDef.interfaces ?? []) {
          const supertype = getNamedType(supertypeDef.name.value)
          setKind(supertype, "object", { fields: [], subtypes: [], supertypes: [] })
          type.supertypes.push(supertype)
          supertype.subtypes.push(type)
        }
      }
    } else if (typeDef.kind === "UnionTypeDefinition" || typeDef.kind === "UnionTypeExtension") {
      setKind(type, "object", { fields: [], subtypes: [], supertypes: [] })
      for (const subtypeDef of typeDef.types ?? []) {
        const subtype = getNamedType(subtypeDef.name.value)
        setKind(subtype, "object", { fields: [], subtypes: [], supertypes: [] })
        type.subtypes.push(subtype)
        subtype.supertypes.push(type)
      }
    } else if (typeDef.kind === "EnumTypeDefinition" || typeDef.kind === "EnumTypeExtension") {
      setKind(type, "enum", { values: [] })
      for (const value of typeDef.values ?? []) {
        type.values.push({
          name: value.name.value,
          directives: value.directives?.slice() ?? [],
        })
      }
    } else {
      types.delete(name)
    }
  }

  return [...types.values()]

  function getNamedType(name: string): NamedType {
    const existing = types.get(name)
    if (existing !== undefined) {
      return existing
    }
    const value = { kind: "unresolved", name, directives: [] } as any as NamedType
    types.set(name, value)
    return value
  }

  function getType(node: TypeNode): Type {
    let nullable = true
    if (node.kind === "NonNullType") {
      node = node.type
      nullable = false
    }
    let type: Type
    if (node.kind === "NamedType") {
      type = getNamedType(node.name.value)
    } else {
      type = { kind: "array", inner: getType(node.type) }
    }
    if (nullable) {
      type = { kind: "nullable", inner: type }
    }
    return type
  }

  function setKind<K extends NamedType["kind"]>(
    type: NamedType | UnresolvedType,
    kind: K,
    init: Omit<Extract<NamedType, { kind: K }>, keyof UnresolvedType>,
  ): asserts type is Extract<NamedType, { kind: K }> {
    if ((type.kind === "unresolved") as boolean) {
      type.kind = kind
      Object.assign(type, init)
    }
    if (type.kind !== kind) throw new Error("Conflicting definitions for type " + type.name)
  }

  function getNamedTypeOfType(type: Type): NamedType {
    while (type.kind === "nullable" || type.kind === "array") {
      type = type.inner
    }
    return type
  }
}

export { Context }
