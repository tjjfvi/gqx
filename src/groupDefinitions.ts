import { Context } from ".";
import { DefinitionNode } from "graphql";

export default (ctx: Context, definitions: readonly DefinitionNode[]) => {
  definitions.map(def => {
    if(def.kind === "InputObjectTypeDefinition" || def.kind === "InputObjectTypeExtension")
      ctx.inputTypes[def.name.value] = (ctx.inputTypes[def.name.value] || []).concat(def.fields);
    else if(def.kind === "EnumTypeDefinition")
      ctx.enumTypes[def.name.value] = (ctx.enumTypes[def.name.value] || []).concat(def.values);
    else if(def.kind === "ObjectTypeDefinition" || def.kind === "ObjectTypeExtension")
      ctx.objectTypes[def.name.value] = (ctx.objectTypes[def.name.value] || []).concat(def.fields);
  });
}
