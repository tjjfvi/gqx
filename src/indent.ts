
function indent(str: string): string;
function indent(strs: TemplateStringsArray, ...interps: string[]): string;
function indent(strs: string | TemplateStringsArray, ...interps: string[]): string{
  const str = typeof strs === "string" ? strs : strs.map((s, i) => s + (interps[i] || "")).join("");
  return str.replace(/^(?=.)/gm, "  ");
}

export default indent;
