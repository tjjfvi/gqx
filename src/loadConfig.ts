import yaml from "js-yaml";

interface Config {
  vars: { [x: string]: [string, boolean] };
  types: { [x: string]: [string, boolean] };
}

const regex = /\/\* @gqx\b([^]+?)\*\/|\/\/ @gqx( .+)?/;

export default (template: string): [Config, (x: string) => string] => {
  const match = template.match(regex);
  if(!match)
    throw new Error("No @gqx comment found");
  const str = (match[1] || match[2] || "").trim() || "{}";
  const config = validateConfig(yaml.safeLoad(str));
  return [config, x => template.replace(regex, () => x).replace(/^\/\/ @ts-nocheck/, "")];
}

const validateConfig = (_config: unknown): Config => {
  if(typeof _config !== "object")
    throw new Error("Config must be an object");
  const config: Config = { vars: {}, types: {} };
  const { vars: _vars = {}, types: _types = {} } = (_config as { vars?: unknown; types?: unknown });
  [["vars", _vars], ["types", _types]].map(([name, v]: ["vars"|"types", unknown]) => {
    if(typeof v !== "object")
      throw new Error(`Config.${name} must be an object`);
    Object.entries(v).map(([k, v]: [string, unknown]) => {
      if(typeof v !== "string" && !(v instanceof Array))
        throw new Error(`Config.${name}[string] must be of type: string | [string] | [string, boolean]`);
      let [a, b = true] = typeof v === "string" ? [v] : (v as unknown[]);
      if(typeof a !== "string")
        throw new Error(`Config.${name}[string][0] must be of type: string`);
      if(typeof b !== "boolean" && typeof b !== "undefined")
        throw new Error(`Config.${name}[string][1] must be of type: ?boolean`);
      config[name][k] = [a, b];
    })
  });
  return config;
}

export { Config, validateConfig };
