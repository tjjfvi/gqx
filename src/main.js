(function (deps, factory) {
    if (typeof module === 'object' && typeof module.exports === 'object') {
        var v = factory(require, exports); if (v !== undefined) module.exports = v;
    }
    else if (typeof define === 'function' && define.amd) {
        define(deps, factory);
    }
})(["require", "exports", "graphql", "."], function (require, exports) {
    var graphql_1 = require("graphql");
    var _1 = require(".");
    exports["default"] = function (_a) {
        var schema = _a.schema, template = _a.template;
        var ast = graphql_1.parse(schema);
        var config = JSON.parse(template.match(/\/\/ @gqx (.+)/)[1]);
        // @ts-ignore
        var schemaDef = ast.definitions.find(function (d) { return d.kind === "SchemaDefinition"; });
        var operations = schemaDef.operationTypes.map(function (o) { return [o.operation, o.type.name.value]; });
        var ctx = {
            operations: operations,
            baseTypes: operations.map(function (o) { return o[1]; }),
            enumTypes: {},
            inputTypes: {},
            objectTypes: {},
            exports: []
        };
        ast.definitions.map(function (def) {
            if (def.kind === "InputObjectTypeDefinition" || def.kind === "InputObjectTypeExtension")
                ctx.inputTypes[def.name.value] = (ctx.inputTypes[def.name.value] || []).concat(def.fields);
            else if (def.kind === "EnumTypeDefinition")
                ctx.enumTypes[def.name.value] = def;
            else if (def.kind === "ObjectTypeDefinition" || def.kind === "ObjectTypeExtension")
                ctx.objectTypes[def.name.value] = (ctx.objectTypes[def.name.value] || []).concat(def.fields);
        });
        var code = [
            _1.stringifyEnumTypes(ctx),
            _1.stringifyInputTypes(ctx),
            _1.stringifyObjectTypes(ctx),
            _1.stringifyExports(ctx),
        ].join("\n\n");
        return code;
    };
});
