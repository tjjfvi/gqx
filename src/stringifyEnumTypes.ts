
import { Context } from ".";

export default (ctx: Context) =>
  Object.entries(ctx.enumTypes).map(([name, values]) => {
    ctx.exports.push(name);
    return `type ${name} =${values.map(v => `\n  | ` + JSON.stringify(v.name.value)).join("")}`
  }).join("\n")
