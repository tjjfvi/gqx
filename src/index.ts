
import {
  parse,
  SchemaDefinitionNode,
  EnumTypeDefinitionNode,
  InputValueDefinitionNode,
  FieldDefinitionNode,
  EnumValueDefinitionNode
} from "graphql";
import stringifyEnumTypes from "./stringifyEnumTypes";
import stringifyInputTypes from "./stringifyInputTypes";
import stringifyObjectTypes from "./stringifyObjectTypes";
import stringifyExports from "./stringifyExports";
import stringifyOperations from "./stringifyOperations";
import stringifyBoilerplate from "./stringifyBoilerplate";

interface Context {
  exports: string[];
  inputTypes: { [k: string]: InputValueDefinitionNode[] };
  enumTypes: { [k: string]: EnumValueDefinitionNode[] };
  objectTypes: { [k: string]: FieldDefinitionNode[] };
  operations: [string, string][];
  baseTypes: string[];
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  config: any;
}

export default ({ schema, template }: { schema: string; template: string }) => {

  const ast = parse(schema);
  const config = JSON.parse(template.match(/\/\/ @gqx (.+)/)[1]);

  // @ts-ignore
  const schemaDef: SchemaDefinitionNode = ast.definitions.find(d => d.kind === "SchemaDefinition");
  const operations: [string, string][] = schemaDef.operationTypes.map(o => [o.operation, o.type.name.value]);
  const ctx: Context = {
    operations,
    baseTypes: operations.map(o => o[1]),
    enumTypes: {},
    inputTypes: {},
    objectTypes: {},
    exports: [],
    config,
  };

  const code = [
    stringifyBoilerplate(),
    stringifyEnumTypes(ctx),
    stringifyInputTypes(ctx),
    stringifyObjectTypes(ctx),
    stringifyOperations(ctx),
    stringifyExports(ctx),
  ].join("\n\n").replace(/\n\n\n+/g, "\n\n");

  return template.replace(/\/\/ @gqx .+/, () => code).replace(/^\/\/ @ts-nocheck/, "");

}

export { Context };
