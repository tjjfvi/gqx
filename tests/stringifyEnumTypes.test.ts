
import { stringifyEnumTypes } from "../src/stringifyEnumTypes";

test("", () => {
  const result = stringifyEnumTypes(utils.getOperations(["stringifyEnumTypes"]));
  expect(result).toParse().toMatchSnapshot();
})
