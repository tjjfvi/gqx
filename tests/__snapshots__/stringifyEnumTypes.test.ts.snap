// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
"type ABC =
  | \\"A\\"
  | \\"B\\"
  | \\"C\\"
type OnlyOneOption =
  | \\"JustThis\\"
type Letter =
  | \\"A\\"
  | \\"B\\"
  | \\"C\\"
  | \\"D\\"
  | \\"E\\"
  | \\"F\\"
  | \\"G\\"
  | \\"H\\"
  | \\"I\\"
  | \\"J\\"
  | \\"K\\"
  | \\"L\\"
  | \\"M\\"
  | \\"N\\"
  | \\"O\\"
  | \\"P\\"
  | \\"Q\\"
  | \\"R\\"
  | \\"S\\"
  | \\"T\\"
  | \\"U\\"
  | \\"V\\"
  | \\"W\\"
  | \\"X\\"
  | \\"Y\\"
  | \\"Z\\"
type AlphaNum =
  | \\"C1\\"
  | \\"P2\\"
  | \\"P3\\"
  | \\"C4\\"
  | \\"P5\\"
  | \\"C6\\"
  | \\"P7\\"
  | \\"C8\\"
  | \\"C9\\"
  | \\"C10\\"
type Underscores =
  | \\"_\\"
  | \\"__\\"
  | \\"___\\"
  | \\"____\\"
  | \\"_____\\""
`;
